File: backend/routers/courses.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from backend.database import get_db
from backend.deps import require_profil_owner
from backend.models.profil import Profil
from backend.schemas.liste_course import (
    CoursesTerminerRequest,
    CoursesTerminerResponse,
    ListeCourseItemCreate,
    ListeCourseItemOut,
    ListeCourseItemUpdate,
)
from backend.services import liste_course_service

router = APIRouter(prefix="/planning", tags=["courses"])


@router.get("/{profil_id}/courses/items", response_model=list[ListeCourseItemOut])
def list_course_items(
    profil: Profil = Depends(require_profil_owner),
    include_done: bool = False,
    db: Session = Depends(get_db),
):
    return liste_course_service.list_items(db, profil.id, include_done=include_done)


@router.post(
    "/{profil_id}/courses/items",
    response_model=ListeCourseItemOut,
    status_code=201,
)
def create_course_item(
    payload: ListeCourseItemCreate,
    profil: Profil = Depends(require_profil_owner),
    db: Session = Depends(get_db),
):
    return liste_course_service.create_item(db, profil.id, payload)


@router.patch(
    "/{profil_id}/courses/items/{item_id}",
    response_model=ListeCourseItemOut,
)
def update_course_item(
    item_id: str,
    payload: ListeCourseItemUpdate,
    profil: Profil = Depends(require_profil_owner),
    db: Session = Depends(get_db),
):
    try:
        return liste_course_service.update_item(db, profil.id, item_id, payload)
    except ValueError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc


@router.delete("/{profil_id}/courses/items/{item_id}", status_code=204)
def delete_course_item(
    item_id: str,
    profil: Profil = Depends(require_profil_owner),
    db: Session = Depends(get_db),
):
    try:
        liste_course_service.delete_item(db, profil.id, item_id)
    except ValueError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc


@router.post(
    "/{profil_id}/courses/terminer",
    response_model=CoursesTerminerResponse,
)
def terminer_courses(
    payload: CoursesTerminerRequest,
    profil: Profil = Depends(require_profil_owner),
    db: Session = Depends(get_db),
):
    try:
        return liste_course_service.terminer_courses(
            db, profil.id, payload.item_ids or None, label=payload.label
        )
    except ValueError as exc:
        raise HTTPException(status_code=422, detail=str(exc)) from exc

File: frontend/src/api/courses.ts
import { api } from "./http";

export type ListeCourseItem = {
  id: string;
  profil_id: string;
  ingredient_id: string | null;
  label: string;
  quantite: number;
  unite: string;
  prix_estime: number | null;
  coche: boolean;
  custom: boolean;
  done: boolean;
};

export function listCourseItems(
  profilId: string,
  token: string,
  includeDone = false
) {
  const q = includeDone ? "?include_done=true" : "";
  return api<ListeCourseItem[]>(`/planning/${profilId}/courses/items${q}`, {
    token,
  });
}

export function createCourseItem(
  profilId: string,
  token: string,
  body: {
    label: string;
    ingredient_id?: string | null;
    quantite?: number;
    unite?: string;
    prix_estime?: number | null;
    custom?: boolean;
  }
) {
  return api<ListeCourseItem>(`/planning/${profilId}/courses/items`, {
    method: "POST",
    token,
    body,
  });
}

export function updateCourseItem(
  profilId: string,
  token: string,
  itemId: string,
  body: Partial<{
    label: string;
    quantite: number;
    unite: string;
    prix_estime: number | null;
    coche: boolean;
  }>
) {
  return api<ListeCourseItem>(
    `/planning/${profilId}/courses/items/${itemId}`,
    { method: "PATCH", token, body }
  );
}

export function deleteCourseItem(
  profilId: string,
  token: string,
  itemId: string
) {
  return api<void>(`/planning/${profilId}/courses/items/${itemId}`, {
    method: "DELETE",
    token,
  });
}

export function terminerCourses(
  profilId: string,
  token: string,
  itemIds: string[],
  label = "Courses"
) {
  return api<{
    items_termines: number;
    stock_approvisionne: number;
    montant_restant: number;
  }>(`/planning/${profilId}/courses/terminer`, {
    method: "POST",
    token,
    body: { item_ids: itemIds, label },
  });
}
